### PR TITLE
draft: support passing Secret source over MCP

### PR DESCRIPTION
**Pull request by doujiang24**
It's a draft for #40325 , and using k8s `core.v1.Secret` according to @howardjohn 's suggestion.

TODO:
1. need to authorize secret resources, simply skip for now, since the current authorize way depends on k8s, but the original intention is to bypass k8s; I haven't figured out a way for it yet.
2. tests and doc.

@howardjohn @hzxuzhonghu  Could you please take a look? Just want to make sure it's in the right direction.
Any feedbacks are welcome, thanks!

### CODE REVIEW

**Comment by doujiang24** [comment]
Hi team, gentle ping.
Do you think it's the right way to pass `v1.Secret` through MCP?

cc @howardjohn @hzxuzhonghu

File: pilot/pkg/xds/sds.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ import (
 	"crypto/x509"
 	"encoding/pem"
 	"fmt"
+	"istio.io/istio/pilot/pkg/credentials/kube"
```
Comment (zirain) [inline pilot/pkg/xds/sds.go:21]: please run `make gen & make lint`

**Review by hzxuzhonghu — COMMENTED** [review]
Will it influence configcontroller with mcp

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -504,14 +504,17 @@ func (s *Server) WaitUntilCompletion() {
 
 // initSDSServer starts the SDS server
 func (s *Server) initSDSServer() {
-	if s.kubeClient == nil {
+	if s.kubeClient == nil && !features.EnableSecretOverMCP {
 		return
 	}
-	if !features.EnableXDSIdentityCheck {
+	if !features.EnableXDSIdentityCheck && !features.EnableSecretOverMCP {
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/server.go:510]: we should still check this for MCP I think? I think you mentioned that in PR description though

File: pilot/pkg/model/push_context.go [hunk 3]
```diff
@@ -197,6 +198,9 @@ type PushContext struct {
 	// gatewayIndex is the index of gateways.
 	gatewayIndex gatewayIndex
 
+	// secrets for each namespace
+	SecretsByNameSpace map[string]map[string]config.Config
```
Comment (howardjohn) [inline pilot/pkg/model/push_context.go:202]: what are the keys? namespace -> name -> config?

File: pilot/pkg/model/push_context.go [hunk 4]
```diff
@@ -1265,6 +1276,12 @@ func (ps *PushContext) updateContext(
 		ps.ServiceAccounts = oldPushContext.ServiceAccounts
 	}
 
+	if secretChanged {
+		if err := ps.initSecrets(env); err != nil {
+			return err
+		}
+	}
```
Comment (howardjohn) [inline pilot/pkg/model/push_context.go:1283]: do we need `else { .. copy old ones ... }`?

File: pilot/pkg/model/push_context.go [hunk 5]
```diff
@@ -1416,6 +1433,22 @@ func (ps *PushContext) initServiceRegistry(env *Environment) error {
 	return nil
 }
 
+// pre computes secrets per namespace
+func (ps *PushContext) initSecrets(env *Environment) error {
+	secretConfigs, err := env.List(gvk.Secret, NamespaceAll)
```
Comment (howardjohn) [inline pilot/pkg/model/push_context.go:1438]: Won't this get all secrets from all components? Including k8s?

File: pilot/pkg/model/push_context.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/model/push_context.go:202]: should we just store a struct of NamespacedName?

Comment (howardjohn) [reply pilot/pkg/model/push_context.go:202]: and `*config.Config` possibly to reduce copies?

File: pilot/pkg/model/push_context.go (hunk 5 above)
Comment (doujiang24) [reply pilot/pkg/model/push_context.go:1438]: I think, no. It get configs from `model.ConfigStore`. but, seems the secrets from k8s, is on another path, `credentials.Controller`, which is not a kind of `model.ConfigStore`.

**Comment by doujiang24** [comment]
> Will it influence configcontroller with mcp

@hzxuzhonghu Thanks.
I think it depends. when EnableSecretOverMCP = true,
1. if means only use secrets from MCP, I think they are two independent ways, they won't influence each other,
2. if means using secrets from k8s and MCP together, then we have to decide the priority, k8s first? or MCP first.

In the current implementation, we don't use secrets from k8s and MCP together, only use one.

File: pilot/pkg/xds/sds.go (hunk 1 above)
Comment (doujiang24) [reply pilot/pkg/xds/sds.go]: Thanks, addressed the style issue. but I got such an error to run `make gen`.

```
bin/update_crds.sh
Cloning into '/tmp/tmp.FlyhketY9O'...
The authenticity of host 'github.com (140.82.114.3)' can't be established.
ED25519 key fingerprint is SHA256:+DiY3wvvV6TuJJhbpZisF/zLDA0zPMSvHdkr4UvCOqU.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Could not create directory '/home/user/.ssh' (No such file or directory).
Failed to add the host to the list of known hosts (/home/user/.ssh/known_hosts).
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.
```

Seems it's something unexpected in the base docker image, will take a deep look later.

**Comment by hzxuzhonghu** [comment]
If we intend to use secret over MCP to provide gateway certs, it is a little bit complex. Even worse in multi cluster, we have to respect both secret namespace and cluster attribute.

From the original issue, I guess you want to use it to provide gateway cert. So it is just a first step for transport,  not for usage

**Comment by doujiang24** [comment]
Thanks. All addressed.

I have two questions, looking forward to your help, thanks.

1. for the secrets, provide a new option way to bypass k8s totally, is the right way? or do you guys think it is worth maintaining it?

background: we are migrating a large & old north-south ingress, from Nginx to Istio + Envoy. And the old control plane is not in the k8s system. So, we hope the new Istio way does not rely on k8s, at least for the first step.
Also, we are processing well in our internal fork, in the bypass k8s way.

2. is it worth introducing a new secret controller?

As said in this TODO comment, so that we may register customer secret authorize method in the feature.
https://github.com/istio/istio/pull/40509/commits/8a21eaa253502dadc260db714a19cf7c92eb3b82#diff-91e0a18a58ea36d2c656d7b2706c52917a74103820c07fc843f48a110c847678R536

We just passing a `nil` secret controller now, which makes it difficult to add `authorize` for secrets.

**Comment by doujiang24** [comment]
> If we intend to use secret over MCP to provide gateway certs, it is a little bit complex. Even worse in multi cluster, we have to respect both secret namespace and cluster attribute.
> 
> From the original issue, I guess you want to use it to provide gateway cert. So it is just a first step for transport, not for usage

Yeah, we want to pass secrets through MCP, so that the data plane could get the gateway cert from istio, and the related name of the gateway cert is defined in `credentialName` in `Gateway` CRD.
https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings

For the usage:
the current implementation (in this PR) does only honor the namespace and secret name, for now.
Missing the `authorize` method, that the biggest left issue for this PR, I think.

**Comment by costinm** [comment]
I am concerned with this - we need a VERY strong security review/discussion first, messing with secrets is dangerous. 

Also: Istiod already has a mechanism to distribute secrets using SDS ( which is a first class XDS object ), with a relatively reasonable security model we discussed at length. I don't know if k8s Secret and envoy Secret are a good match - I don't 
mind exposing k8s Secret, but I think it must have the same security model, the caller must have some explicit permission for the secret.

**Comment by costinm** [comment]
On a related note - we currently use the special generator for MCP, originally it was intended to keep things clean and it was an experiment, but I think it is very reasonable to allow 'MCP' generators to be used with the default (envoy) and envoy resources to be returned on the MCP generator. 

Only 'special' generator is proxyless - mainly because it requires a different format for LDS/RDS, but it also reuses EDS/CDS.

**Comment by doujiang24** [comment]
Sorry, I think I made a mistake in the title, "passing" should be "accepting".

For istiod, we still deliver secret resources over SDS, just like it does now.
One thing that changed is istiod also accepts k8s secret from MCP server,
just like it accept k8s secret from k8s server.

Backgroud: we want to get rid of k8s server, in our north-south gateway.

Yeah, I totally agree with you about security.
I think we may introduce a new secret controller to implement the `authurize`.
But I'm not sure about it, as said in:
https://github.com/istio/istio/pull/40509#issuecomment-1225130704

Need your help, thanks.

**Comment by doujiang24** [comment]
Thanks for your reply.

Sorry, I think I made a mistake in the title, "passing" should be "accepting".

For istiod, we still deliver secret resources over SDS, from istiod to envoy, just like it does now.
eg. using `credentialName` in the `Gateway` CRD, (https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings)

One thing that changed is istiod also accepts k8s secret from MCP server,
just like it accept k8s secret from k8s server.

Backgroud: we want to get rid of k8s server, in our north-south gateway.

Yeah, I totally agree with you about security.
I think we may introduce a new secret controller to implement the `authurize`.
But I'm not sure about it, as said in:
https://github.com/istio/istio/pull/40509#issuecomment-1225130704

Need your help, thanks.

**Pull request closed without merging**